src/cpp/session/modules/SessionCodeSearch.cpp

```diff
@@ -1543,6 +1552,99 @@ SourceItem fromCppDefinition(const clang::CppDefinition& cppDefinition)
       safe_convert::numberTo<int>(cppDefinition.location.column, 1));
 }
 
+void fillFromBookdownRefs(const std::string& term,
+                          std::vector<SourceItem>* pSourceItems)
+{
+   // retrieve refs for this project
+   core::json::Value bookdownIndex = module_context::bookdownXRefIndex();
+   
+   // may be null if we have no bookdown refs (typically implies
+   // we're not in a bookdown project)
+   if (!bookdownIndex.isObject())
+      return;
+   
+   std::string baseDir;
+   core::json::Array bookdownRefs;
+   
+   Error error = core::json::readObject(
+            bookdownIndex.getObject(),
+            "baseDir", baseDir,
+            "refs", bookdownRefs);
+   
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   FilePath basePath = module_context::resolveAliasedPath(baseDir);
+   
+   // {
+   //     "file": "01-intro.Rmd",
+   //     "type": "h1",
+   //     "id": "intro",
+   //     "title": "Introduction"
+   // }
+   
+   for (const json::Value& bookdownRef : bookdownRefs)
+   {
+      if (!bookdownRef.isObject())
+         continue;
+      
+      std::string file, type, id, title;
+      Error error = core::json::readObject(
+               bookdownRef.getObject(),
+               "file", file,
+               "type", type,
+               "id", id,
+               "title", title);
+      
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+      
+      // figure out appropriate source item type
+      SourceItem::Type sourceType = SourceItem::None;
+      if (type == "fig")
+      {
+         sourceType = SourceItem::Figure;
+      }
+      else if (type == "tab")
+      {
+         sourceType = SourceItem::Table;
+      }
+      else if (type == "h1" || type == "h2" || type == "h3" ||
+               type == "h4" || type == "h5" || type == "h6")
+      {
+         sourceType = SourceItem::Section;
+      }
+      
+      // form appropriate text for display
+      std::string displayText = title;
```

Comment:
I'm torn as to whether we should be showing IDs here, at least for figures, tables, and equations -- as that's the 'identifier' that's being searched for (i.e. for functions we don't show the roxygen title if available we show the name of the function). I've also seen that figure and table captions can get quite long (even multiple sentences) and I think the id would be a better mental "bookmark' in the case cases.

For section I can see the argument that we should use the actual title (since the id is typically algorithmically derived from the title).

Reply:
How about both, like `title (#id)` ? Or is that likely to be too long?

Reply:
I think it will be too long (a section title would typically be ~ 10-30 characters which is already over the effective limit for the UI).

Reply:
If we think displaying the ID is necessary we could truncate the section title and always append the ID.

Reply:
I'd say try it without, and then if it's frustrating we figure out how to add it.

Reply:
I think no ID for sections/headers to start with is fine, but we definitely should use the IDs for figures, tables, and equations. In those cases that's the actual "symbol" they are navigating to (it's the chunk label).

Reply:
Sorry, to re-iterate: do we want to be able to search based on the contents of the caption (e.g. for a figure), or only based on the figure ID?

Reply:
I think we should search based on contents, but we should show the id in the list